File: java/PathSumII.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**

 Given a binary tree and a sum, find all root-to-leaf paths where each path's sum equals the given sum.

 For example:
 Given the below binary tree and sum = 22,
    5
    / \
    4   8
    /   / \
    11  13  4
    /  \    / \
     7    2  5   1
 returnhttps://leetcode.com/problems/path-sum-ii/discuss/
 [
 [5,4,11,2],
 [5,8,4,5]
 ]

 */
public class PathSumII {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> result = new LinkedList<>();

        if (root == null) {
            return result;
        }

        dfs(root, 0, sum, new LinkedList<>(), result);

        return result;
    }

    public void dfs(TreeNode root,
                    int curSum,
                    int target,
                    List<Integer> curList,
                    List<List<Integer>> result) {
        if (root == null) {
            return;
        }

        curSum = curSum + root.val;
        curList.add(root.val);

        if (root.left == null && root.right == null && curSum == target) {
            result.add(new ArrayList<>(curList));
        } else {
            dfs(root.left, curSum, target, curList, result);
            dfs(root.right, curSum, target, curList, result);
        }

        //System.out.println(Arrays.toString(curList.toArray()));

        curSum = curSum - root.val;
        curList.remove(curList.size() - 1);
    }
}

File: Tree/path-sum/path-sum/path-sum-ii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  const result = [];

  recursion(root, targetSum, [], result);

  return result;
}

function recursion(
  node: TreeNode,
  targetSum: number,
  current: number[],
  result: number[][]
) {
  if (node === null) {
    return;
  }

  if (node.left === null && node.right === null && node.val === targetSum) {
    result.push([...current, node.val]);
    return;
  }

  current.push(node.val);

  recursion(node.left, targetSum - node.val, current, result);
  recursion(node.right, targetSum - node.val, current, result);

  current.pop();
}
